heareval/score.py

```diff
@@ -223,6 +225,100 @@ class EventBasedScore(SoundEventScore):
+        assert predictions.ndim == 2
+        assert targets.ndim == 2  # One hot
+        auc = roc_auc_score(targets, predictions, average=None)
+        d_prime = stats.norm().ppf(auc) * np.sqrt(2.0)
+        # Averaged over the classes
+        return np.mean(d_prime)
+
+
+class LWLrap(ScoreFunction):
+    """
+    LWLRAP is calculated per class followed by averaging this class level score
+    as mentioned in the below issue
+    https://github.com/neuralaudio/hear2021-secret-tasks/issues/26
+
+    Adopted from -
+    https://colab.research.google.com/drive/1AgPdhSp7ttY18O3fEoHOQKlt_3HJDLi8#scrollTo=FJv0Rtqfsu3X # noqa: E501
+    Author: Dan Ellis dpwe@google.com
+    Date: 2019-03-03
+    """
+
+    name = "lwlrap"
+
+    @staticmethod
+    def _one_sample_positive_class_precisions(scores, truth):
+        """Calculate precisions for each true class for a single sample."""
+        num_classes = scores.shape[0]
+        pos_class_indices = np.flatnonzero(truth > 0)
+        if not len(pos_class_indices):
+            return pos_class_indices, np.zeros(0)
+        retrieved_classes = np.argsort(scores)[::-1]
+        class_rankings = np.zeros(num_classes, dtype=np.int)
+        class_rankings[retrieved_classes] = range(num_classes)
+        retrieved_class_true = np.zeros(num_classes, dtype=np.bool)
+        retrieved_class_true[class_rankings[pos_class_indices]] = True
+        retrieved_cumulative_hits = np.cumsum(retrieved_class_true)
+        precision_at_hits = retrieved_cumulative_hits[
+            class_rankings[pos_class_indices]
+        ] / (1 + class_rankings[pos_class_indices].astype(np.float))
+        return pos_class_indices, precision_at_hits
+
+    def calculate_per_class_lwlrap(self, truth, scores):
+        """Calculate label-weighted label-ranking average precision."""
+        assert truth.shape == scores.shape
+        num_samples, num_classes = scores.shape
+        precisions_for_samples_by_classes = np.zeros((num_samples, num_classes))
+        for sample_num in range(num_samples):
+            (
+                pos_class_indices,
+                precision_at_hits,
+            ) = self._one_sample_positive_class_precisions(
+                scores[sample_num, :], truth[sample_num, :]
+            )
+            precisions_for_samples_by_classes[
+                sample_num, pos_class_indices
+            ] = precision_at_hits
+        labels_per_class = np.sum(truth > 0, axis=0)
+        per_class_lwlrap = np.sum(
+            precisions_for_samples_by_classes, axis=0
+        ) / np.maximum(1, labels_per_class)
+        return per_class_lwlrap
+
+    def __call__(self, predictions: np.ndarray, targets: np.ndarray, **kwargs) -> float:
+        assert predictions.ndim == 2
+        assert targets.ndim == 2  # One hot
+        per_class_lwlrap = self.calculate_per_class_lwlrap(targets, predictions)
+        return np.mean(per_class_lwlrap)
```

Comment:
How does this differ than the google colab? Are we macro averaging as opposed to the weighted averaging? Is that what Eduardo suggested for lwlwrap? Maybe make a quick comment above about what has changed? Also, do we have a reference for lwlwrap any where -- it's hard to tell what this is doing by looking at the code, a brief description and reference would be awesome